### PR TITLE
feat(meetings): Hash trees implementation for Locus syncs

### PR DESCRIPTION
**Pull request by marcin-bazyl**
# COMPLETES #[SPARK-697765](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-697765)

## This pull request addresses
Locus is using hash trees for sending DTO updates in 5k webinars.

## by making the following changes
The work for making web client work with Locus hash tree DTOs was done on a branch visible in this PR:
https://github.com/webex/webex-js-sdk/pull/4427

In order to merge these changes into next branch, above changes were split into 2 PRs:
1. https://github.com/webex/webex-js-sdk/pull/4583 - focuses on connecting the existing SDK LocusInfo code with new HashTreeParser
2. This PR - implements HashTreeParser and HashTree classes

So, this PR depends on PR#4583 and in terms of production code, it has the same code as PR#4427, but has more unit tests.

HashTree class implementation was taken from Colin's initial implementation done a few months ago and modified to use a different library for the hashing algorithm plus some bug fixes done.

### Change Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [ ] Internal code refactor

## The following scenarios where tested

unit tests, e2e manual testing of code from PR#4427 with Locus and web app

## The GAI Coding Policy And Copyright Annotation Best Practices ##
  
- [ ] GAI was not used (or, no additional notation is required)
- [ ] Code was generated entirely by GAI
- [ ] GAI was used to create a draft that was subsequently customized or modified
- [x] Coder created a draft manually that was non-substantively modified by GAI (e.g., refactoring was performed by GAI on manually written code)
- [ ] Tool used for AI assistance (GitHub Copilot / Other - specify)
  - [x] Github Copilot
  - [ ] Other - Please Specify
- [ ] This PR is related to
  - [ ] Feature
  - [x] Defect fix
  - [ ] Tech Debt
  - [ ] Automation
  
### I certified that

- [x] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [x] I discussed changes with code owners prior to submitting this pull request

- [x] I have not skipped any automated checks
- [x] All existing and new tests passed
- [x] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: packages/@webex/plugin-meetings/src/locus-info/index.ts [hunk 1]
```diff
@@ -478,12 +480,19 @@ export default class LocusInfo extends EventsScope {
...
+        );
+        // continuing as we can still manage without responseBody.dataSets, but this is very suspicious
+      }
       LoggerProxy.logger.info(
-        'Locus-info:index#handleLocusAPIResponse: skipping handling of API http response with hashTreeParser'
+        'Locus-info:index#handleLocusAPIResponse --> passing Locus API response to HashTreeParser: ',
+        responseBody
       );
+      // update the data in our hash trees
+      this.hashTreeParser.handleLocusUpdate(responseBody);
```
Comment (mickelr) [inline packages/@webex/plugin-meetings/src/locus-info/index.ts:495]: Is that necessary to wrap the function call in try/catch to avoid potentially breaking?

Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/locus-info/index.ts:495]: hmm I'm not sure.... the code inside HashTreeParser.handleLocusUpdate() is not supposed to throw in any expected edge case and the existing code for non-hash trees also doesn't catch any exceptions here, so feels like any exception that happens should be just thrown and caught higher up in the code that actually invokes the Locus API for which the response is parsed here. Only that code in the upper layer will know really the full context and what API was called so it can decide what to do and how important the failure is. If I was to catch the exception here, I don't see what I could do with it, all I could do is log it and I would still probably have to re-throw it.

File: packages/@webex/plugin-meetings/src/hashTree/hashTree.ts [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+
+  /**
+   * Constructs a new HashTree.
+   * @param {LeafDataItem[]} leafData Initial data to populate the tree.
+   * @param {number} numLeaves The number of leaf nodes in the tree. Must be 0 or a power of 2.
+   * @throws {Error} If numLeaves is not 0 or a power of 2.
+   */
+  constructor(leafData: LeafDataItem[], numLeaves: number) {
+    // eslint-disable-next-line no-bitwise
+    if ((numLeaves & (numLeaves - 1)) !== 0) {
```
Comment (jor-row) [inline packages/@webex/plugin-meetings/src/hashTree/hashTree.ts:28]: Nice simple check!

File: packages/@webex/plugin-meetings/test/unit/spec/hashTree/hashTree.ts [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+      
+      const diff1_2 = tree1.diffHashes(tree2.getHashes());
+      expect(diff1_2).to.deep.equal([1]); // Leaf 1 should differ
+
+      const diff2_1 = tree2.diffHashes(tree1Hashes);
+      expect(diff2_1).to.deep.equal([1]);
+    });
+  });
+
+  describe('computeTreeHashes', () => {
```
Comment (jor-row) [inline packages/@webex/plugin-meetings/test/unit/spec/hashTree/hashTree.ts:590]: a test case for numLeaves = 0 could be added here

File: packages/@webex/plugin-meetings/src/hashTree/hashTree.ts [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+    return results;
+  }
+
+  /**
+   * Computes the hash for a specific leaf.
+   * The hash is based on the sorted items within the leaf.
+   * @param {number} index The index of the leaf to compute the hash for.
+   * @returns {void}
+   */
+  computeLeafHash(index: number) {
```
Comment (jor-row) [inline packages/@webex/plugin-meetings/src/hashTree/hashTree.ts:240]: Would it be worth checking that the index is a valid lead index at the start of this function?

File: packages/@webex/plugin-meetings/src/hashTree/hashTree.ts [hunk 5]
```diff
@@ -0,0 +1,459 @@
...
+    changedLeafIndexes.forEach((index) => {
+      this.computeLeafHash(index);
+    });
+
+    return results;
+  }
+
+  /**
+   * Updates multiple items in the hash tree.
+   * This method can handle both adding and removing items based on the `operation` flag.
```
Comment (jor-row) [inline packages/@webex/plugin-meetings/src/hashTree/hashTree.ts]: I think this comment should be "both updating and removing"

File: packages/@webex/plugin-meetings/src/hashTree/hashTree.ts [hunk 6]
```diff
@@ -0,0 +1,459 @@
...
+   *
+   * @param {object[]} itemUpdates An array of objects containing `operation` flag and the `item` to update.
+   * @returns {boolean[]} An array of booleans indicating success for each operation.
+   */
+  updateItems(itemUpdates: {operation: 'update' | 'remove'; item: LeafDataItem}[]): boolean[] {
```
Comment (jor-row) [inline packages/@webex/plugin-meetings/src/hashTree/hashTree.ts:203]: Unit tests for this function would be great since it is stubbed out in the tests for hashTreeParser

File: packages/@webex/plugin-meetings/src/hashTree/hashTree.ts [hunk 7]
```diff
@@ -0,0 +1,459 @@
...
+   */
+  getLeafCount(): number {
+    return this.numLeaves;
+  }
+
+  /**
+   * Calculates the total number of items stored in the tree.
+   * @returns {number} The total number of items.
+   */
+  getTotalItemCount(): number {
```
Comment (jor-row) [inline packages/@webex/plugin-meetings/src/hashTree/hashTree.ts:342]: I know this is used in the unit tests quite a bit, but do you think it be worth adding some explicit/"dumb" ```getTotalItemCount``` unit tests?

File: packages/@webex/plugin-meetings/test/unit/spec/hashTree/hashTree.ts
Comment (jor-row) [inline packages/@webex/plugin-meetings/test/unit/spec/hashTree/hashTree.ts:1]: Great test coverage!

File: packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts [hunk 8]
```diff
@@ -117,7 +243,195 @@ class HashTreeParser {
...
+          `HashTreeParser#initializeDataSets --> ${this.debugId} initializing dataset "${name}" (${debugText})`
+        );
+
+        this.dataSets[name] = {
+          ...dataSet,
+        };
+      } else {
+        LoggerProxy.logger.info(
+          `HashTreeParser#initializeDataSets --> ${this.debugId} dataset "${name}" already exists (${debugText})`
+        );
```
Comment (jor-row) [inline packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts:287]: question: just checking my understanding here

so if a dataSet already exists, then a hash tree for that dataSet will already exist as well and will have it's own handling of getting the most up to date information so that is why we do nothing here? There is no need to send an update/sync request to locus because of that

File: packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts [hunk 9]
```diff
@@ -117,7 +243,195 @@ class HashTreeParser {
...
+      if (Array.isArray(currentLocusPart)) {
+        for (const item of currentLocusPart) {
+          findAndStoreMetaData(item);
+        }
+      } else {
+        for (const key of Object.keys(currentLocusPart)) {
+          if (Object.prototype.hasOwnProperty.call(currentLocusPart, key)) {
+            findAndStoreMetaData(currentLocusPart[key]);
+          }
+        }
```
Comment (jor-row) [inline packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts:376]: silly q: can the currentLocusPart only be either an array or an object? Would it be problematic to type it as such? There are technically some other types in js that would satisfy  the ```typeof currentLocusPart !== 'object' || currentLocusPart === null``` check such as a set/map (although I understand we won't receive these kinds of types from locus)

File: packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts [hunk 10]
```diff
@@ -128,7 +442,424 @@ class HashTreeParser {
...
+
+  /**
+   * Adds entries to the passed in updateObjects array
+   * for the changes that result from adding and removing visible data sets.
+   *
+   * @param {HashTreeMessage} message - The hash tree message that triggered the visible data set changes.
+   * @param {string[]} addedDataSets - The list of added data sets.
+   * @returns {Promise<void>}
+   */
+  private async initializeNewVisibleDataSets(
```
Comment (jor-row) [inline packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts:691]: Would it be possible for a dataSet to be removed from the visible list before we get around to doing this? I guess it isn't worth putting a check in here to see if they are still visible because the next call to processVisibleDataSetChanges will clean everything up anyway (just thinking "aloud" here)

File: packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts [hunk 11]
```diff
@@ -128,7 +442,424 @@ class HashTreeParser {
...
+
+    LoggerProxy.logger.info(
+      `HashTreeParser#parseMessage --> ${this.debugId} received message ${debugText || ''}:`,
+      message
+    );
+    if (message.locusStateElements?.length === 0) {
+      LoggerProxy.logger.warn(
+        `HashTreeParser#parseMessage --> ${this.debugId} got empty locusStateElements!!!`
+      );
+      // todo: send a metric
```
Comment (jor-row) [inline packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts:740]: How come we want to send a metric here ? Is this something that you've seen happening in testing that shouldn't? Wondering if there are any other places we should send metrics

File: packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts [hunk 12]
```diff
@@ -139,7 +870,292 @@ class HashTreeParser {
...
+      LoggerProxy.logger.info(
+        `HashTreeParser#runSyncAlgorithm --> ${this.debugId} Data set "${dataSet.name}" has no hash tree, skipping sync algorithm`
+      );
+
+      return;
+    }
+
+    dataSet.hashTree.resize(receivedDataSet.leafCount);
+
+    // temporary log for the workshop // todo: remove
```
Comment (jor-row) [inline packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts:943]: I'm guessing we want to keep these temp logs for debugging?

File: packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts (hunk 8 above)
Comment (jor-row) [reply packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts:287]: Could the size of the dataTree and version etc be out of date until the next message arrives?

File: packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts [hunk 13]
```diff
@@ -128,7 +442,424 @@ class HashTreeParser {
    * @returns {void}
    */
   handleLocusUpdate(update: {dataSets?: Array<DataSet>; locus: any}): void {
-    // todo
+    const {dataSets, locus} = update;
```
Comment (jor-row) [inline packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts:445]: should we be running this dataSets object through updateDataSetInfo here?

File: packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts (hunk 9 above)
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts:376]: as discussed offline, I prefer to keep the type as "any" as we might easily miss something if we try to list all possible types, but at the same time the type will never be any JS data structure like set/map, because it's a Locus DTO that's a result of a call to JSON.parse()

File: packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts (hunk 8 above)
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts:287]: as discussed offline, yes if a hash tree exists, it will be kept up-to-date via the heartbeat mechanism, but there is one things we can add here to handle a case if we never get any message at all: we can setup the sync timer (call `runSyncAlgorithm()`) for each of the data sets here

File: packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts (hunk 11 above)
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts:740]: yes, it was happening during workshop when we had a bug in hash calculation - we were incorrectly thinking that we're out of sync and that triggered Locus to send us a message with a list of out of sync objects, but because we were actually in-sync (only our hash calculation was wrong) that message resulted in being empty. 
There may be some other places we want to send metrics - I've raised SPARK-753139 to track this

File: packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts (hunk 12 above)
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts:943]: yeah, probably while 5k webinars are only testable in integration then it's worth keeping them, we'll remove them later on

File: packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts (hunk 13 above)
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts:445]: yes, I think we should add this here

File: packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts (hunk 10 above)
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts:691]: yeah I guess theoretically it is possible, but the datasets we get from Locus in line 692 are only supposed to be the visible ones at that moment in time (because the url we're using is `visibleDataSetsUrl`) and also as you say it would get cleaned up on next message, so I think we're OK here

File: packages/@webex/plugin-meetings/src/hashTree/hashTree.ts (hunk 2 above)
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/hashTree/hashTree.ts:28]: this line of code and the rest of initial implementation of the hash tree class was done by Colin

File: packages/@webex/plugin-meetings/test/unit/spec/hashTree/hashTree.ts (hunk 3 above)
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/test/unit/spec/hashTree/hashTree.ts:590]: done

File: packages/@webex/plugin-meetings/src/hashTree/hashTree.ts (hunk 4 above)
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/hashTree/hashTree.ts:240]: done

File: packages/@webex/plugin-meetings/src/hashTree/hashTree.ts (hunk 5 above)
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/hashTree/hashTree.ts]: fixed

File: packages/@webex/plugin-meetings/src/hashTree/hashTree.ts (hunk 6 above)
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/hashTree/hashTree.ts:203]: added unit tests

File: packages/@webex/plugin-meetings/src/hashTree/hashTree.ts (hunk 7 above)
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/hashTree/hashTree.ts:342]: added tests

File: packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts (hunk 8 above)
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts:287]: actually, I've found that runSyncAlgorithm() is already called - it's called because we call sendInitializationSyncRequestToLocus() a few lines above and that calls parseMessage() which calls runSyncAlgorithm(), so I've only updated the relevant unit tests to add a check that the timers are set.
There is potentially 1 edge case where the timers won't be set - that's if the sync request we send to Locus in sendInitializationSyncRequestToLocus() fails. I'll check with Locus team what we should do in that case, but let's take this out of scope of this PR.

File: packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts (hunk 13 above)
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/hashTree/hashTreeParser.ts:445]: done - call added and test updated

**Review by jor-row — APPROVED** [review]

**Review by mickelr — APPROVED** [review]
